server/src/main/java/com/cloud/agent/manager/allocator/impl/FirstFitAllocator.java

```diff
@@ -417,6 +419,10 @@ public boolean isVirtualMachineUpgradable(VirtualMachine vm, ServiceOffering off
         // Determine the guest OS category of the template
         String templateGuestOSCategory = getTemplateGuestOSCategory(template);
 
+        if (Strings.isNullOrEmpty(templateGuestOSCategory)) {
```

Comment:
if you return a `null` value in `getTemplateGuestOSCategory` method, why not simply check `templateGuestOSCategory == null`?

Is it possible to have an empty or blank `guestOSCategory.name`? 
BTW: `Strings.isNullOrEmpty` will return `false` for blanks.

Reply:
Technically `templateGuestOSCategory == null` would be enough, and I cannot think of a way `getTemplateGuestOSCategory` returning empty string unless the database record integrity has been changed, that's why I used `Strings.isNullOrEmpty` rather rather than a simple `== null` check. (which might be slightly overkill at the same time!)